File: frontend-docs/src/utils/enlacesDeArchivos.js
// frontend-docs/src/utils/enlacesDeArchivos.js
//
// ENLACES DE ARCHIVOS (14-sep-2026): la obra, la carpeta y el documento van en la
// dirección, como en ACC (`folderUrn` + `entityId`). Todo puro, para probarlo sin
// navegador (`pruebas/enlacesDeArchivos.prueba.mjs`).
//
//   /?obra=<obra>                                        la obra, en su carpeta raíz
//   /?obra=<obra>&carpeta=<carpeta>                      una carpeta
//   /?obra=<obra>&carpeta=<carpeta>&documento=<doc>      un documento, en su versión vigente
//   ...&documento=<doc>&version=<versión>                una versión fija de ese documento
//
// Identificadores, nunca nombres: renombrar o mover no rompe el enlace. La raíz de la
// obra no lleva `carpeta`. Una dirección con `revision` es de Revisiones
// (`utils/revisiones.js`) y aquí no se toca.
//
// LA AUTORIDAD NO ESTÁ AQUÍ. Un enlace no concede nada: dónde está hoy lo que nombra, y
// si quien lo abre puede verlo, lo dice el servidor (`GET /api/docs/ubicacion`).

// Donde espera un enlace mientras se inicia sesión: el login puede limpiar la URL.
export const CLAVE_DEL_ENLACE_DE_ARCHIVOS = 'ecd_enlace_archivos';

// La misma frase que `backend/enlaces_de_archivos.MENSAJE_NO_DISPONIBLE`: inexistente, de
// otra obra, en la papelera o sin permiso se dicen igual, sin nombre ni ubicación.
export const ENLACE_NO_DISPONIBLE = 'No tienes acceso a ese elemento o ya no existe.';

const PARAMETROS = ['obra', 'carpeta', 'documento', 'version'];
const DEL_EXPLORADOR = ['carpeta', 'documento', 'version'];
const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export function esIdentificador(valor) {
  return UUID.test(String(valor ?? '').trim());
}

function cadena(p) {
  const s = p.toString();
  return s ? `?${s}` : '';
}

// El enlace de Archivos de una dirección, o null. Sin documento no hay versión.
export function leerEnlaceDeArchivos(search) {
  const p = new URLSearchParams(search || '');
  if (p.has('revision')) return null;
  const valor = (clave) => (p.get(clave) || '').trim() || null;
  const obra = valor('obra');
  if (!obra) return null;
  const documento = valor('documento');
  return { obra, carpeta: valor('carpeta'), documento, version: documento ? valor('version') : null };
}

// La `search` con este enlace, o sin ninguno (null), conservando lo demás. Quita también
// `revision`: la dirección no puede decir dos cosas a la vez.
export function conArchivos(search, enlace) {
  const p = new URLSearchParams(search || '');
  p.delete('revision');
  for (const clave of PARAMETROS) p.delete(clave);
  if (enlace?.obra) {
    p.set('obra', String(enlace.obra));
    if (enlace.carpeta) p.set('carpeta', String(enlace.carpeta));
    if (enlace.documento) {
      p.set('documento', String(enlace.documento));
      if (enlace.version) p.set('version', String(enlace.version));
    }
  }
  return cadena(p);
}

// Fuera de la vista de carpetas la dirección no dice carpeta ni documento: se quedan la
// obra y lo demás, `revision` incluida.
export function sinCarpetaNiDocumento(search) {
  const p = new URLSearchParams(search || '');
  for (const clave of DEL_EXPLORADOR) p.delete(clave);
  return cadena(p);
}

// «Copiar enlace»: la dirección completa. Por omisión, el documento vigente y no una
// versión congelada.
export function enlaceDeArchivos(origen, enlace) {
  return `${origen || ''}/${conArchivos('', enlace)}`;
}

export function mismoEnlace(a, b) {
  if (!a || !b) return !a && !b;
  return PARAMETROS.every(clave => String(a[clave] ?? '') === String(b[clave] ?? ''));
}

// La ruta del explorador (`obra/Carpeta/Sub/`) a partir de la cadena del servidor. Los
// nombres solo sirven para pintar: el enlace va por identificador.
export function rutaDeLaCadena(prefijo, ruta) {
  const nombres = (ruta || []).map(c => c.name);
  return `${prefijo}/${nombres.length ? `${nombres.join('/')}/` : ''}`;
}

// Lo que se guarda en `history.state` en las entradas de Archivos.
export function estadoDeArchivos(enlace, extra = {}) {
  return { alephia: 'archivos', obra: enlace?.obra ?? null, carpeta: enlace?.carpeta ?? null,
           documento: enlace?.documento ?? null, version: enlace?.version ?? null, ...extra };
}

// ¿Lo que trae Atrás/Adelante saca al explorador de otra sección y lo lleva a la vista de
// carpetas? Solo si nombra una carpeta o un documento de ESTA obra: `?obra=` a secas es
// también la dirección de Revisiones y del resto de la obra.
export function pideLaVistaDeCarpetas(enlace, obraActual) {
  return Boolean(enlace && obraActual != null && String(enlace.obra) === String(obraActual)
                 && (enlace.carpeta || enlace.documento));
}

// ATRÁS Y ADELANTE FUERA DEL EXPLORADOR (`App_Refactor`):
//   · 'enlace': la dirección es de Archivos y de una obra que no se está viendo;
//   · 'lista' / 'hub': la entrada es la de la lista de obras o la de la portada;
//   · 'nada': lo resuelve el explorador, o no es de Archivos.
// La lista y la portada solo se reconocen por su marca en `history.state`: una dirección
// vacía también la escribe Revisiones al cerrar un detalle, dentro de la obra.
export function destinoDeArchivosTrasNavegar(search, estado, { enDocumentos = false, obraActual = null } = {}) {
  const enlace = leerEnlaceDeArchivos(search);
  if (enlace) {
    const dentro = enDocumentos && obraActual != null && String(enlace.obra) === String(obraActual);
    return dentro ? { tipo: 'nada' } : { tipo: 'enlace', enlace };
  }
  if (new URLSearchParams(search || '').has('revision')) return { tipo: 'nada' };
  if (estado?.alephia === 'hub') return { tipo: 'hub' };
  if (estado?.alephia === 'lista') return { tipo: 'lista' };
  return { tipo: 'nada' };
}

File: backend/enlaces_de_archivos.py
# -*- coding: utf-8 -*-
"""ARCHIVOS · DÓNDE ESTÁ HOY LO QUE NOMBRA UN ENLACE INTERNO.

EL ENLACE (docs/archivos/01_ENLACES_POR_CARPETA_Y_DOCUMENTO.md)
----------------------------------------------------------------
    /?obra=<obra>                                      la obra, en su carpeta raíz
    /?obra=<obra>&carpeta=<carpeta>                    una carpeta
    /?obra=<obra>&carpeta=<carpeta>&documento=<doc>    un documento, en su versión vigente
    ...&documento=<doc>&version=<versión>              una versión fija de ese documento

Identificadores, nunca nombres: renombrar o mover no rompe el enlace. Es el mismo
modelo que ACC (`folderUrn` + `entityId`).

UN ENLACE NO CONCEDE NADA
-------------------------
Se valida siempre usuario -> obra -> recurso -> permiso y, para una versión, además
versión -> documento. La obra la comprueba la ruta (`verify_project_access`) antes de
llamar aquí; el permiso es `permiso_documental.permiso_efectivo`, la misma regla que
decide el listado, la vista previa y la descarga.

LA CARPETA DA CONTEXTO; EL DOCUMENTO DA IDENTIDAD
-------------------------------------------------
Con `documento`, la `carpeta` de la dirección no decide nada: manda dónde está hoy el
documento. Si se movió de A a C se devuelve C, y la pantalla corrige la dirección. Un
documento nunca se abre dentro de una carpeta solo porque la dirección lo diga.

RESPUESTA NEUTRA
----------------
Inexistente, de otra obra, en la papelera, oculto o sin permiso: todo es
`NoDisponible`, y la ruta responde lo mismo en todos los casos, sin nombre ni
ubicación. El motivo viaja en la excepción para el registro y las pruebas; nunca sale
al cliente.
"""
import os
import uuid

MENSAJE_NO_DISPONIBLE = 'No tienes acceso a ese elemento o ya no existe.'
CODIGO_NO_DISPONIBLE = 'ENLACE_NO_DISPONIBLE'

# El mismo tope con el que `permiso_documental` sube por las carpetas: una cadena más
# larga es un ciclo o un dato roto, y no se sigue.
_MAX_SALTOS = 40

_CLAVES = ('id', 'parent_id', 'name', 'node_type', 'folder_type', 'model_urn',
           'is_deleted', 'status')


class NoDisponible(Exception):
    """El enlace no se abre. `motivo` es para el registro y las pruebas, no para el cliente."""

    def __init__(self, motivo):
        super().__init__(motivo)
        self.motivo = motivo


def _identificador(valor, que):
    """El id tal como lo guarda la base, o None si no viene. Mal formado no abre nada,
    y no llega a consultarse."""
    if valor in (None, ''):
        return None
    try:
        return str(uuid.UUID(str(valor).strip()))
    except (ValueError, TypeError, AttributeError):
        raise NoDisponible('%s mal formado' % que)


def cadena(cur, nodo_id):
    """El nodo y sus carpetas hasta arriba, empezando por el propio nodo."""
    cur.execute("""
        WITH RECURSIVE cadena AS (
            SELECT id, parent_id, name, node_type, folder_type, model_urn, is_deleted,
                   status, 0 AS salto
              FROM file_nodes
             WHERE id::text = %s
            UNION ALL
            SELECT p.id, p.parent_id, p.name, p.node_type, p.folder_type, p.model_urn,
                   p.is_deleted, p.status, c.salto + 1
              FROM file_nodes p
              JOIN cadena c ON p.id = c.parent_id
             WHERE c.salto < %s
        )
        SELECT id::text, parent_id::text, name, node_type, folder_type, model_urn,
               is_deleted, status
          FROM cadena
         ORDER BY salto""", (nodo_id, _MAX_SALTOS))
    return [dict(zip(_CLAVES, fila)) for fila in cur.fetchall()]


def _oculto_por_iso(cur, usuario, model_urn, estado):
    """La regla de `file_system_db.list_contents`: con STRICT_ISO_VISIBILITY, quien no
    administra la obra no ve lo que no está Compartido, Publicado o Archivado. Un enlace
    no puede enseñar lo que el listado esconde."""
    if os.getenv('STRICT_ISO_VISIBILITY', 'false').lower() not in ('true', '1', 'yes'):
        return False
    import estados_ecd as ecd
    try:
        from administracion_de_obra import es_admin_de_obra
        if es_admin_de_obra(cur, usuario, model_urn):
            return False
    except Exception:
        pass                                   # FAIL-CLOSED, como el listado
    return ecd.normalizar(estado) not in {ecd.SHARED, ecd.PUBLISHED, ecd.ARCHIVED}


def ubicar(cur, usuario, model_urn, carpeta=None, documento=None, version=None,
           puede_descargar=None):
    """Dónde está hoy lo que nombra el enlace, si quien pregunta puede verlo.

    Devuelve `{'carpeta', 'ruta', 'documento', 'version'}`:
      carpeta    id de la carpeta que hay que abrir; None si es la raíz de la obra
      ruta       [{'id', 'name'}] de arriba abajo, sin la raíz: con ella la pantalla
                 compone la ruta que ya usa el explorador
      documento  id del documento, o None
      version    la versión pedida con los mismos campos que `/api/docs/versions`, y
                 su clave de almacenamiento solo si `puede_descargar(documento)`; o None

    Cualquier otro caso lanza `NoDisponible`.
    """
    carpeta = _identificador(carpeta, 'carpeta')
    documento = _identificador(documento, 'documento')
    version = _identificador(version, 'version')
    if version and not documento:
        raise NoDisponible('version sin documento')
    objetivo = documento or carpeta
    if not objetivo:
        raise NoDisponible('enlace sin carpeta ni documento')

    # 1 · EL RECURSO, DENTRO DE ESTA OBRA, fuera de la papelera y con toda su cadena.
    filas = cadena(cur, objetivo)
    if not filas:
        raise NoDisponible('no existe')
    if any(f['model_urn'] != model_urn for f in filas):
        raise NoDisponible('de otra obra')
    # `is not False`: el listado pide `is_deleted = FALSE`, así que un nulo tampoco se ve.
    if any(f['is_deleted'] is not False for f in filas):
        raise NoDisponible('en la papelera')
    if filas[-1]['parent_id'] is not None:
        raise NoDisponible('cadena de carpetas rota o demasiado larga')
    nodo = filas[0]
    if nodo['node_type'] != ('FILE' if documento else 'FOLDER'):
        raise NoDisponible('no es del tipo que dice el enlace')
    carpetas = filas[1:] if documento else filas
    if any(f['node_type'] != 'FOLDER' for f in carpetas):
        raise NoDisponible('cadena de carpetas rota')

    # 2 · EL PERMISO EFECTIVO. La raíz de la obra es el enlace de la obra: la ve todo
    # miembro, como en el explorador. Un documento lleva el permiso de su carpeta.
    if documento or nodo['folder_type'] != 'PROJECT_ROOT':
        import permiso_documental as pd
        from folder_permissions import PERMISSION_LEVELS
        nivel = pd.permiso_efectivo(cur, usuario, model_urn, objetivo)
        if PERMISSION_LEVELS.get(nivel, -1) < PERMISSION_LEVELS['viewer']:
            raise NoDisponible('sin permiso')
    if documento and _oculto_por_iso(cur, usuario, model_urn, nodo['status']):
        raise NoDisponible('oculto por el modo ISO estricto')

    # 3 · LA VERSIÓN, DE ESTE DOCUMENTO, antes de devolver nada.
    datos_version = None
    if version:
        cur.execute("SELECT 1 FROM file_versions WHERE id::text = %s AND file_node_id::text = %s",
                    (version, documento))
        if not cur.fetchone():
            raise NoDisponible('la version no es de este documento')
        from file_system_db import get_file_versions
        datos_version = next((v for v in get_file_versions(model_urn, documento)
                              if str(v.get('id')) == version), None)
        if datos_version is None:
            raise NoDisponible('la version no se pudo leer')
        if puede_descargar is None or not puede_descargar(documento):
            datos_version = {k: v for k, v in datos_version.items() if k != 'gcs_urn'}

    visibles = [f for f in carpetas if f['folder_type'] != 'PROJECT_ROOT']
    en_la_raiz = not carpetas or carpetas[0]['folder_type'] == 'PROJECT_ROOT'
    return {
        'carpeta': None if en_la_raiz else carpetas[0]['id'],
        'ruta': [{'id': f['id'], 'name': f['name']} for f in reversed(visibles)],
        'documento': documento,
        'version': datos_version,
    }
